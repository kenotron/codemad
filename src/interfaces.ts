import ts from 'typescript';

export interface Modder {
  replace(node: ts.Node, replacement: string): TypescriptMod;
  append(node: ts.Node, suffix: string): TypescriptMod;
  prepend(node: ts.Node, prefix: string): TypescriptMod;
  remove(node: ts.Node): TypescriptMod;
  removeFull(node: ts.Node): TypescriptMod;
}

export interface Visitor {
  (node: ts.Node, modder: Modder): TypescriptMod | TypescriptMod[] | IterableIterator<TypescriptMod> | undefined | void;
}

export interface TypescriptMod {
  node: ts.Node;
  start: number;
  length: number;
  replacement: string;
}

export interface ModHandlers<T> {
  asJson: (cb: (json: T) => T) => ModHandlers<T>;
  asText: (cb: (text: string) => string) => ModHandlers<T>;
  asTypescript: (visitor: Visitor) => ModHandlers<T>;
  files: ModResult[];
}

export interface ModOptions {
  dryRun?: boolean;
}

export interface ModResult {
  fileName: string;
  state: 'not-modified' | 'would-be-modified' | 'modified';
}
