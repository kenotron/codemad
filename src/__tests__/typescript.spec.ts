import mockfs from 'mock-fs';
import ts from 'typescript';
import { loadFixtures } from './testutils/loadFixtures';
import { testTsHandlerWithFixtureFactory } from './testutils/testTsHandlerWithFixtureFactory';
import { TypescriptMod } from '../interfaces';

const FixturesPath = 'src/__tests__/fixtures/typescript';

describe('tshandler', () => {
  const mockContents = loadFixtures(FixturesPath);
  const testWithFixture = testTsHandlerWithFixtureFactory(FixturesPath);

  beforeEach(() => {
    mockfs(mockContents);
  });

  it('replaces node with text', () => {
    testWithFixture('variableDeclaration.ts', {}, (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.replace(node, `${node.name.getText()} = 'test'`);
      }
    });
  });

  it('prepends node with text', () => {
    testWithFixture('prepend.ts', {}, (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.prepend(node.parent, `console.log('prepended');\n`);
      }
    });
  });

  it('appends node with text', () => {
    testWithFixture('append.ts', {}, (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.append(node.parent, `\nconsole.log('appended');`);
      }
    });
  });

  it('removes node', () => {
    testWithFixture('remove.ts', {}, (node, modder) => {
      if (ts.isExpressionStatement(node)) {
        return modder.remove(node);
      }
    });
  });

  it('removes node fully', () => {
    testWithFixture('removeFull.ts', {}, (node, modder) => {
      if (ts.isExpressionStatement(node)) {
        return modder.removeFull(node);
      }
    });
  });

  it('mods multiple mods in a generator function', () => {
    testWithFixture('iterable.ts', {}, function* (node, modder) {
      if (ts.isVariableDeclaration(node) && node.name.getText() === 'a') {
        yield modder.replace(node.name, 'aaaa');
      } else if (ts.isVariableDeclaration(node) && node.name.getText() === 'c') {
        yield modder.replace(node.name, 'cccc');
      }
    });
  });

  it('mods multiple mods if results in array', () => {
    testWithFixture('array.ts', {}, (node, modder) => {
      const mods: TypescriptMod[] = [];

      if (ts.isVariableDeclaration(node) && node.name.getText() === 'a') {
        mods.push(modder.replace(node.name, 'aaaa'));
      } else if (ts.isVariableDeclaration(node) && node.name.getText() === 'c') {
        mods.push(modder.replace(node.name, 'cccc'));
      }

      return mods;
    });
  });

  it('does not mod when nothing is returned', () => {
    testWithFixture('nomod.ts', {}, (node, modder) => { });
  });


  it("returns list of files modified", () => {
    const results = testWithFixture("variableDeclaration.ts", {}, (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.replace(node, `${node.name.getText()} = 'test'`);
      }
    });
    expect(results[0].fileName).toEqual("src/__tests__/fixtures/typescript/variableDeclaration.ts");
    expect(results[0].state).toEqual("modified");
  });

  it("does not modify files in dry-run mode", () => {
    const results = testWithFixture("dryRun.ts", { dryRun: true }, (node, modder) => {
      if (ts.isVariableDeclaration(node)) {
        return modder.replace(node, `${node.name.getText()} = 'test'`);
      }
    });
    expect(results[0].fileName).toEqual("src/__tests__/fixtures/typescript/dryRun.ts");
    expect(results[0].state).toEqual("modified");
  });
});
