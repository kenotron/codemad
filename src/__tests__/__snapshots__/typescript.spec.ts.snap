// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tshandler appends node with text 1`] = `
"// Here's a comment
var a = 1
console.log('appended');;

var b = 2
console.log('appended');;

var c = 3
console.log('appended');;
"
`;

exports[`tshandler does not mod when nothing is returned 1`] = `
"// Here's a comment
var a = 1;

var b = 2;

console.log('remove me fully');

var c = 3;
"
`;

exports[`tshandler does not modify files in dry-run mode 1`] = `
"// No changes, in dry run.
var a = 1;

var b = 2;

var c = 3;

console.log(a);

console.log(b);
"
`;

exports[`tshandler mods multiple mods if results in array 1`] = `
"// Here's a comment
var aaaa = 1;

var b = 2;

console.log('remove me fully');

var cccc = 3;
"
`;

exports[`tshandler mods multiple mods in a generator function 1`] = `
"// Here's a comment
var aaaa = 1;

var b = 2;

console.log('remove me fully');

var cccc = 3;
"
`;

exports[`tshandler prepends node with text 1`] = `
"// Here's a comment
console.log('prepended');
var a = 1;

console.log('prepended');
var b = 2;

console.log('prepended');
var c = 3;
"
`;

exports[`tshandler removes node 1`] = `
"// Here's a comment
var a = 1;

var b = 2;



var c = 3;
"
`;

exports[`tshandler removes node fully 1`] = `
"// Here's a comment
var a = 1;

var b = 2;

var c = 3;
"
`;

exports[`tshandler replaces node with text 1`] = `
"// Here's a comment
var a = 'test';

var b = 'test';

var c = 'test';

console.log(a);

console.log(b);
"
`;

exports[`tshandler returns list of files modified 1`] = `
"// Here's a comment
var a = 'test';

var b = 'test';

var c = 'test';

console.log(a);

console.log(b);
"
`;
